src/hotspot/share/runtime/arguments.cpp

```diff
@@ -1856,6 +1856,23 @@ void Arguments::set_heap_size() {
       reasonable_max = MAX2(reasonable_max, (julong)MinHeapSize);
     }
 
+    // With current TSan memory mapping of 48bit aarch64, it's very hard to allocate large continuous space.
+    // Add a limitation to avoid the allocation failure in machine with huge physical RAM.
+    // Does not support 39bit and 42bit aarch64 right now. Will add it if necessary.
+    TSAN_ONLY(
+#if defined(AARCH64)
```

Comment:
Could the code be moved to os_linux_aarch64.cpp, by overriding os::has_allocatable_memory_limit()?
There is "reasonable_max = limit_by_allocatable_memory(reasonable_max);" above that calls os::has_allocatable_memory_limit().

We should avoid adding architecture-specific code in arguments.cpp if possible.

Reply:
yes, I think it's possible, however it will touch X86 code that moving has_allocatable_memory_limit() to os_linux_x86.cpp as well.

Reply:
I will move it and test x86. thank you for reminding.

Reply:
@caoman I think it's better to put the code into os_posix.cpp with TSAN_ONLY, AARCH64_ONLY, not os_linux_aarch64.cpp, because latter has to make multiple copies of function has_allocatable_memory_limit for archtectures. And in os_linux.cpp, AARCH64, X86_64 specific code is allowed. what do you think?

Reply:
Thanks for explaining. Let's keep using TSAN_ONLY.